### DISCUSSION TITLE
Cross referencing step parameters

### DISCUSSION THREAD

**Discussion by sphemakh** [SpheMakh/Stimela#697] [Ideas]
I already have this implemented in the [usecwl branch](https://github.com/ratt-ru/Stimela/tree/usecwl) [RecipeStep.py](https://github.com/ratt-ru/Stimela/blob/usecwl/stimela/RecipeStep.py) file. This allows something like this

https://github.com/ratt-ru/Stimela/blob/8563da2e831a36bcd3c6164324e1632a18751616/examples/simulation_pipeline.py#L5-L28

**Comment by o-smirnov** [comment]
You could also do it with the same OmegaConf mechanism that allows config files to cross-reference each other:

* Define a schema for a recipe 

* Dynamically load recipe configs into the global config object under ``recipe.RECIPE_NAME.blah.blah``, as they are being defined

* Refer to other parameters as ``"{recipe.RECIPE_NAME.STEP.PARAMETER}"``

* As a bonus, you can refer to stuff from the base and cabs files as ``"{base.IMAGE.blah.blah}"``

I kind of like it because then we have a unified, hierarchical namespace for everything: configuration, image and cab definitions, recipe definitions, etc.

Also, from this, it's a trivial step to support recipes that are entirely defined as a YaML file.

**Comment by sphemakh** [comment]
> I kind of like it because then we have a unified, hierarchical namespace for everything: configuration, image and cab definitions, recipe definitions, etc. 

I had something like this before but abandoned it because it was not easy to keep track of everything properly. OmegaConf does make this simpler

This is what's left of the old implementation. The idea was to allow a pipeline to be (re-)executed from JSon file

https://github.com/ratt-ru/Stimela/blob/eba6a69b953161d53e973d65f1d6fa74b923aeb4/stimela/recipe.py#L618-L635

**Comment by sphemakh** [comment]
> You could also do it with the same OmegaConf mechanism that allows config files to cross-reference each other: 

Can you point me to where you do this?

**Reply by o-smirnov** [reply]
I don't think I do in the current codebase, but substituting ``{x.y.z}`` as a value is a standard OmegaConf feature, just try it in the current YaML files.

**Comment by o-smirnov** [comment]
Your ``stimela runcab`` (which indeed ought to be ``stimela exec`` I think!) gives me an idea. OmegaConf can be told to load a config section up [from command-line arguments](https://omegaconf.readthedocs.io/en/2.0_branch/usage.html#from-command-line-arguments). You can combine this with ``exec`` very simply to run a cab (i.e. essentially a single-step recipe) from the command line.

**Reply by sphemakh** [reply]
Hehe, I'm actually implementing this at the moment (inspired by your YAML based recipe idea earlier) but did not know OmegaConf can load options from the command-line (currently using argparse).

**Reply by o-smirnov** [reply]
Well you still want to use argparse for everything else, right?  There's a very simple and clean way to combine the two. You add an extra argument to the ``exec`` subparser like so:

```python
subparser.add_argument('params', metavar='NAME=VALUE', type=str, nargs='*', help='cab parameters')
```

...so that everything is collected into args.params. Then create an omegaconf object [using from_dotlist()](https://omegaconf.readthedocs.io/en/2.0_branch/usage.html#from-a-dot-list):

```python
cabconf = OmegaConf.from_dotlist(args.params)
```

...and then, assuming you've gone the route of keeping recipe definitions in the global namespace under ``recipe.NAME`` as we discussed above, you can load it into ``recipe.tmp`` as:

```python
conf = OmegaConf.merge(conf, dict(recipe=dict(tmp=cabconf)))
```

**Reply by sphemakh** [reply]
This doesn't feel right 

```python
subparser.add_argument('params', metavar='NAME=VALUE', type=str, nargs='*', help='cab parameters')
```

I prefer:

```python
for name, value in cab.inputs.items():
    subparser.add_argument(f"--{name}", default=value['default'], help=value.info)
```

**Reply by sphemakh** [reply]
But the `dolist` can be easily be created from the `parser.parser_args()` even in this case

**Reply by o-smirnov** [reply]
Ah ok, so you even build a subparser from the cab definition. OK, that's perhaps even cleaner (and gives ``--help`` for free).

> But the dolist can be easily be created from the parser.parser_args() even in this case

Yes, because you want the value strings to be parsed as YML and checked against the schema, right. Feeding them in to ``from_dotlist()`` will accomplish that (and give you {}-substitutions in the bargain).

**Reply by o-smirnov** [reply]
@SpheMakh there's one potential boo-boo with your approach. The ArgumentParser options namespace is completely flat (even when sub-parsers are in use), so we'll eventually have a name clash between a cab parameter and a Stimela command-line option defined by the parent parser. 

Not sure how to work around this cleanly yet, but it's another reason to reconsider the ``key=value`` approach (which also saves you the trouble of mapping cab parameter types to and from ArgumentParser-recognized types).

**Reply by sphemakh** [reply]
That is not very clever of argparse, it should have a name space per subparser. The `key=value` is a bit ugly

**Reply by sphemakh** [reply]
We can just keep the parsers independent for running a cab on the command line.

**Reply by o-smirnov** [reply]
It's just an assignment statement @SpheMakh, how ugly can it be really. :)

**Reply by sphemakh** [reply]
It's parsing options to a python script on the command-line,  `--foo/-foo` is what is I expect

**Reply by sphemakh** [reply]
We can also deal with the `sys.argv` string directly to resolve this argprase namespace issue

**Reply by sphemakh** [reply]
Somthing like : 

```
cab_args = cab_args_subparser.parse_args(<only cab args>) 
```

**Reply by o-smirnov** [reply]
Yeah but the top-level parser is still going to try to parse the entire ``sys.argv`` list and see a conflict...

If we're looking at doing sys.argv gymnastics just to conform to dogma, it may be time to reconsider the dogma.

> It's parsing options to a python script on the command-line, --foo/-foo is what is I expect

There's a category difference between passing command-line options to a tool, and setting the config parameters of a recipe. If both can be handled via ``--foo`` that's fine, but if this creates problems (as it seems to at the moment), nobody is going to see ``foo=VALUE`` and go "oh my god, what an unnatural syntax, what a shameless pervert that @SpheMakh is".

There's plenty precedent for using ``foo=VALUE``, too. I know ``make`` and ``cmake`` off the top of my head, but there's surely others. And I'm sure they adopted it for a similar reason -- they have a large namespace of [essentially, user-defined] variables coming from the makefiles.

Final advantage: you can let OmegaConf parse ``VALUE`` as per YaML rules. No need to explicitly map between argparse types and internal types.

**Reply by sphemakh** [reply]
Some clever people have already improved on argparse's shortcomings. It's called [click](https://click.palletsprojects.com/en/7.x/api) I particularly like this blurp

## Why not Argparse?

Click is internally based on optparse instead of argparse. This is an implementation detail that a user does not have to be concerned with. Click is not based on argparse because it has some behaviors that make handling arbitrary command line interfaces hard:

argparse has built-in behavior to guess if something is an argument or an option. This becomes a problem when dealing with incomplete command lines; the behaviour becomes unpredictable without full knowledge of a command line. This goes against Click’s ambitions of dispatching to subparsers.

argparse does not support disabling interspersed arguments. Without this feature, it’s not possible to safely implement Click’s nested parsing.

**Reply by sphemakh** [reply]
Have a look at [this example](https://dev.to/drcloudycoder/develop-python-cli-with-subcommands-using-click-4892), it's a much better fit to your restructuring of the `/commands` codebase.

**Reply by o-smirnov** [reply]
Hehe, you'll go a long way to stick up for your dogma, I appreciate that. :P

Click looks very neat though, I vote we take it on board either way.

But here's another feature you can't have with argparse or click (at least easily) -- overriding recipe variables from the command line. ``stimela exec recipe my_selfcal_recipe.yml inputs.ms=foo.ms vars.size=1024 vars.scale=1arcsec``.

**Reply by sphemakh** [reply]
> Hehe, you'll go a long way to stick up for your dogma, I appreciate that. :P 

![dogma](https://www.denofgeek.com/wp-content/uploads/2020/02/tauchertDogma_buddy_christ.jpg)

**Reply by sphemakh** [reply]
> But here's another feature you can't have with argparse or click (at least easily) -- overriding recipe variables from the command line. stimela exec recipe my_selfcal_recipe.yml inputs.ms=foo.ms vars.size=1024 vars.scale=1arcsec.

Agreed. Let's not break our backs over this. Let's go for `foo=VALUE`

**Reply by o-smirnov** [reply]
An excellent choice! You will not regret it.

Also, in this case we can simplify the invocation further by auto-detecting if the first argument to exec is a recipe file or a cab name. ``stimela exec [recipe.yml|cabname] foo=VALUE ...``.

**Reply by o-smirnov** [reply]
Yep, click is neat as fsck. I pushed some changes -- take a look at build or exxec, for example. Everything is so much more succint.

**Reply by o-smirnov** [reply]
While we're at it, should this command not be unified with ``stimela run recipe.py``?

**Reply by sphemakh** [reply]
I thought it was neater to separate recipes that are scripts and those that are yaml files, but there is no fundamental difference between the two; both are just recipes.
